### PR TITLE
Tidy up aarch64 feature detection code in armcap.c

### LINKED ISSUE #20188: Arm feature detection on Apple platforms should use sysctlbyname
As discussed in #14838, the SIGILL-based feature detection that OpenSSL uses on Arm is incorrect and causes problems. On Apple platforms, the API to use is `sysctlbyname`:
https://developer.apple.com/documentation/kernel/1387446-sysctlbyname/determining_instruction_set_characteristics

OpenSSL broadly uses `sysctlbyname`, but still runs the probing mechanism on Apple platforms. It just happens to filter out the not-always-available features on this line:
https://github.com/openssl/openssl/blob/aa2d7e0ee15d1b7015479c38f370a25ceec690fc/crypto/armcap.c#L345

But then later PRs regressed this by not similarly guarding such features: #17916 and #15361. I ran into this because OpenSSL's regression here seems to interfere with debugging things on macOS with `lldb`, but the SIGILL mechanism is also not thread-safe.

But that limited `__APPLE__` guard is also the wrong way to do this. Instead one should completely disable the SIGILL mechanism on platforms where a real API is available. Indeed, as it's not thread-safe, best not to have it at all and just use the correct APIs on all platforms, combined with the preprocessor defines from [ACLE](https://github.com/ARM-software/acle) to handle always-available features since older Apple platforms didn't provide runtime detection for such features. (BoringSSL handles all its detection without SIGILL.)

### PR DESCRIPTION
**Pull request by oraluben**
Cherry-picked from 52a38144b019cfda6b0e5eaa0aca88ae11661a26.

Fixes #20188 for version 3.1.

The preliminary probe is not in 3.0 yet, so there's no need to cherry-pick this "fix" yet.
If applying all the changes in 3.0 is another topic.

Also: https://github.com/openssl/openssl/issues/20188#issuecomment-1545396778

### CODE REVIEW

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM (but I am the author of the original change!)

**Comment by tom-cosgrove-arm** [comment]
Not sure if this should be `triaged: bug` or `triaged: feature`, but pretty sure it counts as `severity: fips change`

**Comment by paulidale** [comment]
OTC: **should this be back ported to 3.1 (& 3.0?)**

**Comment by h-vetinari** [comment]
>@paulidale:  Is any of this relevant to 3.0 as well? It's not entirely clear from the discussion.

@oraluben posted an analysis in the python [issue](https://github.com/python/cpython/issues/99079):

> tldr: now, and hopefully only openssl 3.1.0 is affected.
> 
> Some instruction probing logic breaks debugging in darwin-aarch64, these commits are in the dev branch of openssl (`master`, SVE: https://github.com/openssl/openssl/commit/b1b2146ded9ce5a84c62f30c6c4a922b449f6c90, RNG: https://github.com/openssl/openssl/commit/efa1f22483ee43d84e1aee01b08c0bda04060c1c), and have been cherry-picked to 3.1 release branch (`openssl-3.1`, SVE: https://github.com/openssl/openssl/commit/a14eff6319ec254901e051d61e93d3272c451ebe, RNG: https://github.com/openssl/openssl/commit/45c74dea). They haven't been cherry-picked to 3.0, so openssl 3.0.x is not affected.
> 
> The fix (https://github.com/openssl/openssl/commit/52a38144b019cfda6b0e5eaa0aca88ae11661a26) is also in `master`, I submitted a PR to cherry-pick it in 3.1 (#20948), if it's merged, the only affected version would be `3.1.0`. I guess `3.0.x` would be ok forever, if openssl decided to apply those instruction tests I expect they won't forget the corresponding fix.

**Comment by t8m** [comment]
**OTC: This is major refactoring of the CPU feature detection code. We are not inclined to merge this to 3.1 unless this is having a major impact.**
